assets/js/components/ProjectTextEditor.js

```diff
@@ -0,0 +1,237 @@
+    this.editText.maxLength = config.maxLength
+    this.snackbar = config.snackbar
+    this.closedCallback = closedCallback
+    $('#edit-headline').text(config.headline)
+    $('#edit-close-button').attr('title', config.closeText)
+    $('#edit-text-description').text(config.description)
+    if (config.showLanguageSelect) {
+      $('#edit-language-selector').removeClass('d-none')
+    } else {
+      $('#edit-language-selector').addClass('d-none')
+    }
+
+    self.editTextUI.removeClass('d-none')
+  }
+
+  // region private
+  this.close = () => {
+    if (self.areChangesSaved()) {
+      self._close()
+    } else {
+      self.closeEditorDialog.show(closeEditorDialogResult)
+    }
+  }
+
+  this._close = () => {
+    self.editTextUI.addClass('d-none')
+    self.reset()
+    self.closedCallback()
+  }
+
+  this.reset = () => {
+    self.languageSelect.selectedIndex = 0
+    self.previousIndex = 0
+  }
+
+  function getLanguages () {
+    $.ajax({
+      url: '../languages',
+      type: 'get',
+      success: function (data) {
+        self.languages = data
+        self.populateSelector()
+      }
+    })
+  }
+
+  this.populateSelector = () => {
+    self.languageSelectorList.empty()
+    self.languageSelectorList.append('<li class="mdc-list-item" data-value="default" role="option" tabindex="-1">' +
+      '<span class="mdc-list-item__ripple"></span>' +
+      '<span class="mdc-list-item__text">' + self.defaultText + '</span>' +
+      '</li>')
+
+    for (const language in self.languages) {
+      if (language.length <= 2) {
+        self.languageSelectorList.append(`<li class="mdc-list-item" data-value="${language}" role="option" tabindex="-1">\
+          <span class="mdc-list-item__ripple"></span>\
+          <span class="mdc-list-item__text">${self.languages[language]}</span>\
+          </li>`)
+      }
+    }
+
+    self.languageSelect.layoutOptions()
+    self.reset()
+  }
+
+  this.areChangesSaved = () => {
+    return self.editText.val().trim() === self.lastSavedText.trim()
+  }
+
+  function keepOrDiscardChangesResult (result) {
+    if (result.isConfirmed) {
+      self.previousIndex = self.languageSelect.selectedIndex
+      self.lastSavedText = self.editText.text().trim()
+    } else if (result.isDenied) {
+      self.previousIndex = self.languageSelect.selectedIndex
+      self.getCustomTranslation()
+    } else if (result.isDismissed) {
+      self.languageSelect.selectedIndex = self.previousIndex
+    }
+  }
+
+  function closeEditorDialogResult (result) {
+    if (result.isConfirmed) {
+      self.save()
+    } else if (result.isDenied) {
+      self._close()
+    }
+  }
+
+  this.save = () => {
+    const newText = self.editText.val().trim()
+    if (self.areChangesSaved()) {
+      self._close()
+      return
+    }
+
+    const languageSelected = self.languageSelect.value
+    if (languageSelected === '' || languageSelected === 'default') {
+      const url =
+        self.programSection === 'credit'

```

Comment:
I think it might make it easier to read if this was multiple if statements compared to multiple ternary operators